src/cares_wrap.cc

```diff
@@ -748,8 +748,20 @@ void ChannelWrap::StartTimer() {
   }
   int timeout = timeout_;
   if (timeout == 0) timeout = 1;
-  if (timeout < 0 || timeout > 1000) timeout = 1000;
-  uv_timer_start(timer_handle_, AresTimeout, timeout, timeout);
+  /* -1 implies the default c-ares value, that is 5000, thus checking each
+   * second */
+  if (timeout <= -1) timeout = 1000;
+  if (timeout >= 1000) {
+    int thousands = (timeout / 1000) * 1000;
+    /* Pure seconds should be checked every second (for example 1000, 2000,
+     * 5000) */
```

Comment:
Can you explain why full seconds matter here? What’s special about them?

Reply:
To set the timeout on `uv_timer_start` so we call it at least every second to check if there's a timeout.

So basically if a timeout has full seconds then check every second
If a timeout hasn't full seconds, then check the non-seconds part so we can match (for example 2500 check every 500ms)

But I'm not sure if there's any problem calling it every tick (so we don't care about timeouts on the timer), if there's not then I think it would be the optimal solution

Reply:
I think it would be very weird to have timeout values of `5000` and `4999` to have substantially different behavior, though.

Could we maybe just reduce the timeout cap here? E.g. from `1000` to `500`?

Reply:
Which timeout cap? I'm not sure I understand

Reply:
Previously, the timeout was capped at 1 second. We could just lower that value, right?

Reply:
So setting both `timeout` and `repeat` on `uv_timer_start` or just setting `repeat` to 500ms?

Reply:
@oluan Well … why would we set `timeout` to a lower value than `repeat` here?

Reply:
@addaleax because for example, if `timeout` is `500` and c-ares timeouts some milliseconds after that `timeout`, then we will need to wait for one more `timeout`.

Example:

```js
const dns = require('dns');

const resolver = new dns.Resolver({
    timeout: 5 * 1000,
    tries: 1,
});

resolver.setServers([
    "123.123.123.123"
])

const timeStart = performance.now();
resolver.resolve4("myip.opendns.com", (err, ip) => {
    const timeEnd = performance.now();
    const elapsedMs = (timeEnd - timeStart);
    console.log('elapsed:', elapsedMs);
    if (err) {
        console.error(err);
    } else console.log(ip);
})
```

For ```uv_timer_start(timer_handle_, AresTimeout, timeout, timeout);```
Will sometimes output
```bash
luandevecchi@Luans-Air node % ./node test.js
elapsed: 5503.92962500453
Error: queryA ETIMEOUT myip.opendns.com
    at QueryReqWrap.onresolve [as oncomplete] (node:dns:213:19) {
  errno: undefined,
  code: 'ETIMEOUT',
  syscall: 'queryA',
  hostname: 'myip.opendns.com'
}
```

And sometimes 

```bash
luandevecchi@Luans-Air node % ./node test.js
elapsed: 5005.590124994516
Error: queryA ETIMEOUT myip.opendns.com
    at QueryReqWrap.onresolve [as oncomplete] (node:dns:213:19) {
  errno: undefined,
  code: 'ETIMEOUT',
  syscall: 'queryA',
  hostname: 'myip.opendns.com'
}
```

But ```uv_timer_start(timer_handle_, AresTimeout, 0, timeout);```

Will always output something around

```bash
luandevecchi@Luans-Air node % ./node test.js
elapsed: 5036.493957996368
Error: queryA ETIMEOUT myip.opendns.com
    at QueryReqWrap.onresolve [as oncomplete] (node:dns:213:19) {
  errno: undefined,
  code: 'ETIMEOUT',
  syscall: 'queryA',
  hostname: 'myip.opendns.com'
}
```

Reply:
That's odd, apart from the callback invocation at the immediately following event loop turn `uv_timer_start(timer_handle_, AresTimeout, timeout, timeout);` and `uv_timer_start(timer_handle_, AresTimeout, 0, timeout);` should be equivalent, really. In any case, I'm happy with either -- it won't affect performance all *that* much, I assume.

Reply:
Alright

```cpp
int timeout = timeout_;
if (timeout == 0) timeout = 1;
if (timeout == -1 || timeout >= 1000) timeout = 500;
uv_timer_start(timer_handle_, AresTimeout, 0, timeout);
  ```
  
  With the above code, for example:
  
  `timeout = 2050` will thrown an exception after `2500`
  `timeout = 1100` will throw an exception after `1500`
  `timeout = 2600` will throw an exception after `3000` 
  
  Like this:
  ```cpp
  int timeout = timeout_;
  if (timeout == 0) timeout = 1;
  if (timeout == -1) timeout = 500;
  if (timeout >= 1000) {
    int timeout_h = timeout - (timeout / 1000) * 1000;
    if (timeout_h == 0) timeout = 500;
    else if (timeout_h >= 100) timeout = 100;
    else timeout = 10;
  }
  uv_timer_start(timer_handle_, AresTimeout, 0, timeout);
  ```
  
  `timeout = 2050` will thrown an exception after `~2050`
  `timeout = 1100` will throw an exception after `~1100`
  `timeout = 2600` will throw an exception after `~2600`
  
  (again `5000` and `4999` will have slightly different behaviors)
  
  What do you think?

Reply:
> With the above code, for example:

Fwiw, I think you’d want `if (timeout <= 0 || timeout >= 500) timeout = 500;`

> (again 5000 and 4999 will have slightly different behaviors)

I’m sorry, but I still don’t get what benefits we would get from this discontinuous behavior. With this variant, `4999`, `5000` and `5001` will have *three* separate behaviors, even though the user essentially specified the *same* thing.

Reply:
Alright, I'm happy with setting the `timeout` to `500`.